src/auth/openid-manager.ts

```diff
@@ -0,0 +1,187 @@
+import * as client from 'openid-client';
+import PromiseQueue from '../util/promise-queue';
+import { LocalStorageSync } from '../util/localstorage-proxy';
+
+interface RequestState {
+  verifier: string;
+  challenge: string;
+  state: string;
+}
+
+export type Whoami = Record<string, any>;
+
+const OPENID_SCOPES = 'openid email profile offline_access';
+let openidManagerSingleton: OpenidManager | undefined;
+
+export function getOpenIdManager(): OpenidManager {
+  if (!openidManagerSingleton) {
+    openidManagerSingleton = new OpenidManager();
+  }
+  return openidManagerSingleton;
+}
+
+class OpenidManager {
+  private sync: {
+    openid_tokens: {
+      access_token: string;
+      refresh_token?: string;
+      id_token?: string;
+      expires_at: number;
+    } | null;
+    openid_request_state: RequestState | null;
+  };
+  private config: client.Configuration | null = null;
+  private queue: PromiseQueue = new PromiseQueue();
+  private oldCodes: Set<string> = new Set();
+
+  constructor() {
+    this.sync = LocalStorageSync({
+      openid_tokens: null,
+      openid_request_state: null,
+    });
+  }
+
+  public async getAuthorizationUrl(prompt: 'login' | 'registration'): Promise<URL> {
+    return this.queue.add(async (): Promise<URL> => {
```

Comment:
Anything inside the queue gets executed serially. Just make sure not to await this.queue.add() within the callback, or you'll livelock